wagtail/models/__init__.py

```diff
@@ -386,7 +391,164 @@ class Meta:
+        verbose_name=_("live revision"),
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        editable=False,
+    )
+
+    go_live_at = models.DateTimeField(
+        verbose_name=_("go live date/time"), blank=True, null=True
+    )
+    expire_at = models.DateTimeField(
+        verbose_name=_("expiry date/time"), blank=True, null=True
+    )
+    expired = models.BooleanField(
+        verbose_name=_("expired"), default=False, editable=False
+    )
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def check(cls, **kwargs):
+        return [
+            *super().check(**kwargs),
+            *cls._check_revision_mixin(),
+        ]
+
+    @classmethod
+    def _check_revision_mixin(cls):
+        mro = cls.mro()
+        error = checks.Error(
+            "DraftStateMixin requires RevisionMixin to be applied after DraftStateMixin.",
+            hint="Add RevisionMixin to the model's base classes after DraftStateMixin.",
+            obj=cls,
+            id="wagtail.EXXX",
+        )
+
+        try:
+            if mro.index(RevisionMixin) < mro.index(DraftStateMixin):
+                return [error]
+        except ValueError:
+            return [error]
+
+        return []
+
+    @property
+    def status_string(self):
+        if not self.live:
+            if self.expired:
+                return _("expired")
+            else:
+                return _("draft")
+        else:
+            if self.has_unpublished_changes:
+                return _("live + draft")
+            else:
+                return _("live")
+
+    def publish(
+        self, revision, user=None, changed=True, log_action=True, previous_revision=None
+    ):
+        return PublishRevisionAction(
+            revision,
+            user=user,
+            changed=changed,
+            log_action=log_action,
+            previous_revision=previous_revision,
+        ).execute()
+
+    def unpublish(self, set_expired=False, commit=True, user=None, log_action=True):
+        return UnpublishAction(
+            self,
+            set_expired=set_expired,
+            commit=commit,
+            user=user,
+            log_action=log_action,
+        ).execute()
+
+    def with_content_json(self, content):
+        """
+        Returns a new version of the object with field values updated to reflect changes
+        in the provided ``content`` (which usually comes from a previously-saved revision).
+
+        Certain field values are preserved in order to prevent errors if the returned
+        object is saved, such as ``id``. The following field values are also preserved,
+        as they are considered to be meaningful to the object as a whole, rather than
+        to a specific revision:
+
+        * ``latest_revision``
+        * ``live``
+        * ``has_unpublished_changes``
+        * ``first_published_at``
+
+        If ``TranslatableMixin`` is applied, the following field values are also preserved:
```

Comment:
I think we should delete this section from the docstring (this bit is implemented on ``Page``)

Reply:
This `TranslatableMixin` bit is also implemented in `RevisionMixin`.. should we reword the docstring in here to state something like `in addition to the fields preserved in RevisionMixin, the following fields are also preserved`?